Refine Go: Sampling (#3318)
open-telemetry/opentelemetry.io: content/en/docs/instrumentation/go/sampling.md

---
title: Sampling
weight: 80
---

Sampling is a process that restricts the amount of traces that are generated by
a system. The exact sampler you should use depends on your specific needs, but
in general you should make a decision at the start of a trace, and allow the
sampling decision to propagate to other services.

> A [`Sampler`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#Sampler) can
> to be set on the tracer provider when register it using the
> [`WithSampler`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#WithSampler)
> option, as follows:

Reviewer: ```suggestion
When you registar a [`Sampler`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#Sampler) 
you can set it to the tracer provider using the
[`WithSampler`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#WithSampler)
option as follows:
```

Author: I do not understand this change.

If you want to set a sampler then you have to do it via `WithSampler`

I see that my sentence was a total mess. Changing to

```suggestion
A [`Sampler`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#Sampler) can
be set on the tracer provider using the
[`WithSampler`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#WithSampler)
option, as follows:
```

```go
provider := trace.NewTracerProvider(
    trace.WithSampler(trace.AlwaysSample()),
)
```

[`AlwaysSample`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#AlwaysSample)
and
[`NeverSample`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#NeverSample)

> are fairly self-explanatory. Always means that every trace will be sampled, the
> converse holds as true for Never. When you're getting started, or in a
> development environment, you'll almost always want to use `AlwaysSample`.

Reviewer: ```suggestion
are self-explanatory values. `AlwaysSample` means that every trace is
sampled, while `NeverSample` means that no trace is sampled. When
you're getting started, or in a development environment, use
`AlwaysSample`.
```

Reviewer: Are there production recommended values? When one would use `NeverSample`?

Author: E.g. one someone would like to have it configurable.

Reviewer: Could you add a sentence on when would `NeverSample` be useful?

Other samplers include:

- [`TraceIDRatioBased`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#TraceIDRatioBased),
  which samples a fraction of traces, based on the fraction given to the
  sampler. If you set .5, half of all the traces are sampled.

> - [`ParentBased`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#ParentBased),

Reviewer 2: Should ParentBased include something about delegating to another sampler?

Author: PTAL 5847bd3c

  which behaves differently based on the incoming sampling decision.

>   `ParentBased` samples spans that have parents that were sampled, and doesn't

Reviewer 2: nit: can we be consistent about whether the "trace" or the "span" is sampled?  Generally, I would actually prefer "spans" being sampled, but we appear to use "trace is sampled" more often.

Author: https://github.com/open-telemetry/opentelemetry.io/pull/3318/commits/7d941581f6bbfcb3d8247ffac3c2453f96343a01

> - [`TraceIDRatioBased`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#TraceIDRatioBased),
>   which will sample a fraction of traces, based on the fraction given to the
>   sampler. Thus, if you set this to .5, half of traces will be sampled.
> - [`ParentBased`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#ParentBased),
>   which behaves differently based on the incoming sampling decision. In general,
>   this will sample spans that have parents that were sampled, and will not
>   sample spans whose parents were _not_ sampled.

Reviewer: ```suggestion
- [`TraceIDRatioBased`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#TraceIDRatioBased),
  which samples a fraction of traces, based on the fraction given to the
  sampler. If you set .5, half of all the traces are sampled.
- [`ParentBased`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#ParentBased),
  which behaves differently based on the incoming sampling decision. `ParentBased`
  samples spans that have parents that were sampled, and doesn't sample spans
  whose parents were not sampled.
```

By default, the tracer provider uses a `ParentBased` sampler with the

> `AlwaysSample`.

Reviewer: ```suggestion
`AlwaysSample` setting.
```

Author: ```suggestion
`AlwaysSample` sampler.
```

> When you're in production, you should consider using the `ParentBased` sampler

Reviewer: ```suggestion
When in a production environment, consider using the `ParentBased` sampler
```

---
title: Sampling
weight: 80
---

> Sampling is a process that restricts the amount of spans that are generated by a

Reviewer 3: Recommend linking to the concept page, `/docs/concepts/sampling/`

Author: https://github.com/open-telemetry/opentelemetry.io/pull/3318/commits/dc2a8087756660335c32a696436bbe6836a466b5

system. The exact sampler you should use depends on your specific needs, but in

> general you should make a decision at the start of a span, and allow the

Reviewer 2: "start of a trace" was correct here. Alternative: "first span in a trace"

sampling decision to propagate to other services.

A [`Sampler`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#Sampler) can
be set on the tracer provider using the
[`WithSampler`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#WithSampler)
option, as follows:

```go
provider := trace.NewTracerProvider(
    trace.WithSampler(trace.AlwaysSample()),
)
```

[`AlwaysSample`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#AlwaysSample)
and
[`NeverSample`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#NeverSample)
are self-explanatory values. `AlwaysSample` means that every span is sampled,
while `NeverSample` means that no span is sampled. When you're getting started,

> or in a development environment, use `AlwaysSample`.

Reviewer 3: Since this is the default, I think it's worth calling out that you don't need to specify the sampler unless you're already configuring a head sampler and want to change it for development.

Author: The default is different and described at the end.

Other samplers include:

- [`TraceIDRatioBased`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#TraceIDRatioBased),

>   which samples a fraction of spans, based on the fraction given to the sampler.

Reviewer 3: Mmmm, I think we still do want to say traces, or at least clarify what's happening here. Yes, it will sample spans, but it's really about sampling traces at  given rate at the head. For example, if you emit 100 traces each request and specify a rate of 0.5, you won't end up with 50 missing spans in each trace, but you will end up with 50% of the total span volume.

Author: > dif you emit 100 traces each request and specify a rate of 0.5, you won't end up with 50 missing spans in each trace

I am not sure if you are correct. I think this will be correct if you combine `TraceIDRatioBased` and `ParentBased`.

Author: Also see https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#traceidratiobased

  If you set .5, half of all the spans are sampled.
- [`ParentBased`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#ParentBased),
  is a sampler decorator which behaves differently, based on the parent of the
  span. If the span has no parent, the decorated sampler is used to make
  sampling decision based on the parent of the span. By default, `ParentBased`
  samples spans that have parents that were sampled, and doesn't sample spans
  whose parents were not sampled.

By default, the tracer provider uses a `ParentBased` sampler with the
`AlwaysSample` sampler.

When in a production environment, consider using the `ParentBased` sampler with

> the `TraceIDRatioBased` sampler.

Reviewer 3: I'm thinking we may want to revise this, since it's a little complicated.

If you combine both, then yes, it will typically behave how you like. However, if your application is called by an upstream one you don't necessarily control, then it can be advantageous to omit the `ParentBased` composite sampler. The result is that you can sometimes end up with traces that have missing root spans, but the rest of your traces will be coherent and at the volume you care about. We've run into this several times with customers who do head sampling and it unfortunately means they have to pick one or the other. We've tried to get them to switch to tail-based sampling but they often can't.

I'm thinking of rolling that guidance out across all the languages, but if you agree, would you want to incorporate here, or handle orthogonally?

Author: I am not sure if I follow. The `ParentBased` is configurable regarding local vs remote parent spans. See the available options https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#ParentBasedSamplerOption

> I'm thinking of rolling that guidance out across all the languages, but if you agree, would you want to incorporate here, or handle orthogonally?

We can try incorporating it here. If we will have problems doing so then we can handle it orthogonally.

Reviewer 3: Okay, sounds good. We can tackle separately.
